source/custom/xnetsocket_converters.h

```diff
@@ -412,6 +452,23 @@ struct SockOptDataOutputConverter {
         output.mutable_data_linger()->set_l_onoff(data_linger.l_onoff);
         break;
       }
+      case OptName::OPT_NAME_IP_ADD_MEMBERSHIP:
+      case OptName::OPT_NAME_IP_DROP_MEMBERSHIP: {
+        output.mutable_data_ipmreq()->set_imr_multiaddr(data_ipmreq.imr_multiaddr.addr);
+        output.mutable_data_ipmreq()->set_imr_interface(data_ipmreq.imr_interface.addr);
+        break;
+      }
+      case OptName::OPT_NAME_IPV6_ADD_MEMBERSHIP:
+      case OptName::OPT_NAME_IPV6_DROP_MEMBERSHIP: {
+        // Reinterpret unsigned char to char.
+        auto multiaddr_out = reinterpret_cast<const char*>(data_ipv6mreq.ipv6mr_multiaddr.addr);
+        auto multiaddr_size = sizeof(data_ipv6mreq.ipv6mr_multiaddr.addr);
+        output.mutable_data_ipv6mreq()->set_ipv6mr_multiaddr(
+            {&multiaddr_out[0],
+             &multiaddr_out[multiaddr_size]});
```

Comment:
I'm not sure if I have the syntax right but are the arguments supposed to be `(pointer_to_size, size)`? If that's the case I think the second argument should change to just be `multiaddr_size`.

Reply:
Alex's change introduced `copy_ipv6_addr_to_output` helper that sets the IPv6 output message with the storage data. This is what we use now in all outputs dealing with IPv6Addrs.

Reply:
From the [std::string reference](https://en.cppreference.com/w/cpp/string/basic_string/basic_string): I think @reckenro  is referring to (4). This incantation might work as (6) `[first, last)`, but I'm not sure. Pointers should satisfy the contract of `InputIt`, but I Am Not A Compiler. What we definitely **DO NOT WANT** is (5) (`char*`) because that will only copy up to the first null byte. Our test case should catch that because of the `0x0` byte at the beginning. But you can double-check that by changing this to use only pass in the `char*` and see if you can get it to fail (`EXPECTED "string data" ACTUAL ""`)

Reply:
Oh I see, nevermind. I missed the curly braces and thought those were args to `set_ipv6mr_multiaddr`. Yeah, for strings it probably falls into case (6) where it's `[first, last)` so this works out.